plugin/trino-exchange/src/main/java/io/trino/plugin/exchange/FileSystemExchangeSink.java

```diff
@@ -0,0 +1,373 @@
+                writersMap.values().stream().map(BufferedStorageWriter::finish).collect(toImmutableList())));
+        addSuccessCallback(finishFuture, this::destroy);
+        finishFuture = Futures.transformAsync(
+                finishFuture,
+                ignored -> exchangeStorage.createEmptyFile(outputDirectory.resolve(COMMITTED_MARKER_FILE_NAME)),
+                directExecutor());
+        Futures.addCallback(finishFuture, new FutureCallback<>()
+        {
+            @Override
+            public void onSuccess(Void result)
+            {
+                closed = true;
+            }
+
+            @Override
+            public void onFailure(Throwable ignored)
+            {
+                abort();
+            }
+        }, directExecutor());
+
+        return toCompletableFuture(finishFuture);
+    }
+
+    @Override
+    public synchronized CompletableFuture<Void> abort()
+    {
+        if (closed) {
+            return completedFuture(null);
+        }
+        closed = true;
+
+        ListenableFuture<Void> abortFuture = asVoid(Futures.allAsList(
+                writersMap.values().stream().map(BufferedStorageWriter::abort).collect(toImmutableList())));
+        addSuccessCallback(abortFuture, this::destroy);
+
+        return toCompletableFuture(Futures.transformAsync(
+                abortFuture,
+                ignored -> exchangeStorage.deleteRecursively(outputDirectory),
+                directExecutor()));
+    }
+
+    private void throwIfFailed()
+    {
+        Throwable throwable = failure.get();
+        if (throwable != null) {
+            throwIfUnchecked(throwable);
+            throw new RuntimeException(throwable);
+        }
+    }
+
+    private void destroy()
+    {
+        writersMap.clear();
+        bufferPool.close();
+    }
+
+    @ThreadSafe
+    private static class BufferedStorageWriter
+    {
+        private static final int INSTANCE_SIZE = ClassLayout.parseClass(BufferedStorageWriter.class).instanceSize();
+
+        private final ExchangeStorageWriter storageWriter;
+        private final BufferPool bufferPool;
+        private final AtomicReference<Throwable> failure;
+
+        @GuardedBy("this")
+        private SliceOutput currentBuffer;
+
+        public BufferedStorageWriter(ExchangeStorageWriter storageWriter, BufferPool bufferPool, AtomicReference<Throwable> failure)
+        {
+            this.storageWriter = requireNonNull(storageWriter, "storageWriter is null");
+            this.bufferPool = requireNonNull(bufferPool, "bufferPool is null");
+            this.failure = requireNonNull(failure, "failure is null");
+        }
+
+        public synchronized void write(Slice data)
+        {
+            writeInternal(Slices.wrappedIntArray(data.length()));
+            writeInternal(data);
+        }
+
+        public ListenableFuture<Void> finish()
+        {
+            // no synchronization needed as `write` will not run concurrently with `finish`
```

Comment:
It feels like we still need it to be sure of `currentBuffer` visibility. At least it is not obvious why we do not. Maybe let's just add it to be on the safe side.

Reply:
Yeah true. Changing `abort` and `finish` to synchronized here.